File: apps/mobile/lib/services/appwrite_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:appwrite/models.dart' as models;
import 'package:campus_mesh/appwrite_config.dart';
import 'package:flutter/foundation.dart';

/// Appwrite service singleton for managing Appwrite SDK instances
class AppwriteService {
  factory AppwriteService() {
    // Ensure initialization on every access in release mode
    if (!_instance._initialized) {
      debugPrint('WARNING: AppwriteService accessed before initialization!');
      _instance.init();
    }
    return _instance;
  }
  AppwriteService._internal();
  static final AppwriteService _instance = AppwriteService._internal();

  late Client client;
  late Account account;
  late Databases databases;
  late Storage storage;
  late Realtime realtime;

  bool _initialized = false;
  bool get isInitialized => _initialized;

  /// Initialize Appwrite client and services
  void init() {
    if (_initialized) {
      debugPrint('AppwriteService already initialized');
      return;
    }

    try {
      debugPrint(
        'Initializing Appwrite with endpoint: ${AppwriteConfig.endpoint}',
      );
      debugPrint('Project ID: ${AppwriteConfig.projectId}');

      client = Client()
          .setEndpoint(AppwriteConfig.endpoint)
          .setProject(AppwriteConfig.projectId)
          .setSelfSigned(status: false); // Ensure HTTPS is properly validated

      account = Account(client);
      databases = Databases(client);
      storage = Storage(client);
      realtime = Realtime(client);

      _initialized = true;
      debugPrint('✅ Appwrite initialized successfully');
    } catch (e, stackTrace) {
      debugPrint('❌ Failed to initialize Appwrite: $e');
      debugPrint('Stack trace: $stackTrace');
      throw Exception('Appwrite initialization failed: $e');
    }
  }

  /// Check if user is authenticated
  Future<bool> isAuthenticated() async {
    try {
      await account.get();
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Get current user session
  Future<models.User?> getCurrentUser() async {
    try {
      return await account.get();
    } catch (e) {
      return null;
    }
  }
}

File: apps/web/src/services/appwrite.service.ts
/**
 * Appwrite Service
 * Handles all Appwrite client operations
 */

import { Client, Account, Databases, Storage, Query } from 'appwrite';
import { AppwriteConfig } from '../config/appwrite';

class AppwriteService {
  private client: Client;
  public account: Account;
  public databases: Databases;
  public storage: Storage;

  constructor() {
    this.client = new Client()
      .setEndpoint(AppwriteConfig.endpoint)
      .setProject(AppwriteConfig.projectId);

    this.account = new Account(this.client);
    this.databases = new Databases(this.client);
    this.storage = new Storage(this.client);
  }

  /**
   * Get the current authenticated user
   */
  async getCurrentUser() {
    try {
      return await this.account.get();
    } catch (error) {
      console.error('Error getting current user:', error);
      throw error;
    }
  }

  /**
   * Sign in with email and password
   */
  async signIn(email: string, password: string) {
    try {
      return await this.account.createEmailPasswordSession(email, password);
    } catch (error) {
      console.error('Error signing in:', error);
      throw error;
    }
  }

  /**
   * Sign out
   */
  async signOut() {
    try {
      await this.account.deleteSession('current');
    } catch (error) {
      console.error('Error signing out:', error);
      throw error;
    }
  }

  /**
   * Check if user is authenticated
   */
  async isAuthenticated(): Promise<boolean> {
    try {
      await this.account.get();
      return true;
    } catch {
      return false;
    }
  }
}

// Export singleton instance
export const appwriteService = new AppwriteService();
export { Query };
